### PR TITLE
[PM-1807] Add Auth Request Service

### PR DESCRIPTION
**Pull request by justindbaur**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Refactor most of the logic in the `AuthRequestController` into a new `AuthRequestService`. Also add unit tests for the new `AuthRequestService`.

## Code changes

* **src/Api/Auth/Controllers/AuthRequestsController.cs:** Removed all logic that was moved into the new service.
* **src/Core/Auth/Models/Api/Request/AuthRequest/AuthRequestUpdateRequestModel.cs:** Moved request objects to `Core` in the `Auth` team folder.
* **src/Core/Auth/Services/Implementations/AuthRequestService.cs:** New service containing controller logic
* **test/Core.Test/Auth/Services/AuthRequestServiceTests.cs:** Tests containing validation of all service logic.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: test/Core.Test/Auth/Services/AuthRequestServiceTests.cs
Comment (JaredSnider-Bitwarden) [inline test/Core.Test/Auth/Services/AuthRequestServiceTests.cs:1]: Great work on adding these tests!

File: test/Core.Test/Auth/Services/AuthRequestServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            Key = "test_key",
+            DeviceIdentifier = "test_identifier",
+            RequestApproved = true,
+            MasterPasswordHash = "my_hash",
+        };
+
+        // Give it a randomly generated userId such that it won't be valid for the AuthRequest
+        await Assert.ThrowsAsync<NotFoundException>(
+            async () => await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, authRequest.UserId, updateModel));
+    }
```
Comment (JaredSnider-Bitwarden) [inline test/Core.Test/Auth/Services/AuthRequestServiceTests.cs:308]: Here's two additional test cases for the `UpdateAuthRequestAsync` method (untested):

(1) Test case for when _deviceRepository.GetByIdentifierAsync() returns null:
```
[Theory, BitAutoData]
public async Task UpdateAuthRequestAsync_InvalidDeviceIdentifier_ThrowsBadRequest(
    SutProvider<AuthRequestService> sutProvider,
    AuthRequest authRequest)
{
    authRequest.CreationDate = DateTime.UtcNow.AddMinutes(-10);
    authRequest.Approved = null;

    sutProvider.GetDependency<IAuthRequestRepository>()
        .GetByIdAsync(authRequest.Id)
        .Returns(authRequest);

    sutProvider.GetDependency<IDeviceRepository>()
        .GetByIdentifierAsync(Arg.Any<string>(), authRequest.UserId)
        .Returns((Device?)null);

    var updateModel = new AuthRequestUpdateRequestModel
    {
        Key = "test_key",
        DeviceIdentifier = "invalid_identifier",
        RequestApproved = true,
        MasterPasswordHash = "my_hash",
    };

    await Assert.ThrowsAsync<BadRequestException>(
        async () => await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, authRequest.UserId, updateModel));
}

```

(2) Test case for when the `AuthRequest` has already been approved or rejected:
```
[Theory, BitAutoData]
public async Task UpdateAuthRequestAsync_AlreadyApprovedOrRejected_ThrowsDuplicateAuthRequestException(
    SutProvider<AuthRequestService> sutProvider,
    AuthRequest authRequest)
{
    // Set CreationDate to a valid recent value and Approved to a non-null value
    authRequest.CreationDate = DateTime.UtcNow.AddMinutes(-10);
    authRequest.Approved = true;

    sutProvider.GetDependency<IAuthRequestRepository>()
        .GetByIdAsync(authRequest.Id)
        .Returns(authRequest);

    var device = new Device
    {
        Id = Guid.NewGuid(),
        Identifier = "test_identifier",
    };

    sutProvider.GetDependency<IDeviceRepository>()
        .GetByIdentifierAsync(device.Identifier, authRequest.UserId)
        .Returns(device);

    var updateModel = new AuthRequestUpdateRequestModel
    {
        Key = "test_key",
        DeviceIdentifier = "test_identifier",
        RequestApproved = true,
        MasterPasswordHash = "my_hash",
    };

    await Assert.ThrowsAsync<DuplicateAuthRequestException>(
        async () => await sutProvider.Sut.UpdateAuthRequestAsync(authRequest.Id, authRequest.UserId, updateModel));
}
```

**Comment by JaredSnider-Bitwarden** [comment]
Awesome work on this! I left a few minor comments w/ some additional test cases to handle, but otherwise everything seemed very comprehensively tested!

File: test/Core.Test/Auth/Services/AuthRequestServiceTests.cs (hunk 1 above)
Comment (justindbaur) [reply test/Core.Test/Auth/Services/AuthRequestServiceTests.cs:328]: All 3 of these tests were great, I literally only had to bring in a namespace for `DuplicateAuthRequestException` and they all worked perfectly.

Comment (JaredSnider-Bitwarden) [reply test/Core.Test/Auth/Services/AuthRequestServiceTests.cs:328]: Awesome!

**Review by JaredSnider-Bitwarden — APPROVED** [review]
LGTM! Great work on this!

**Comment by justindbaur** [comment]
Merging to master for further testing.

**Pull request merged**